tests/test_blocking_http_server.py

```diff
@@ -0,0 +1,122 @@
+from contextlib import contextmanager
+from copy import deepcopy
+from multiprocessing import Pool
+from urllib.parse import urlparse
+
+import pytest
+import requests
+
+from pytest_httpserver import BlockingHttpServer
+
+
+@pytest.fixture
+def httpserver():
+    server = BlockingHttpServer(timeout=1)
+    server.start()
+
+    yield server
+
+    server.clear()
+    if server.is_running():
+        server.stop()
+
+
+def test_behave_workflow(httpserver: BlockingHttpServer):
+    request = dict(
+        method="GET",
+        url=httpserver.url_for("/my/path"),
+    )
+
+    with when_a_request_is_being_sent_to_the_server(request) as server_connection:
+
+        client_connection = then_the_server_gets_the_request(httpserver, request)
+
+        response = {"foo": "bar"}
+
+        when_the_server_responds_to(client_connection, response)
+
+        then_the_response_is_got_from(server_connection, response)
+
+
+def test_raises_assertion_error_when_request_does_not_match(httpserver: BlockingHttpServer):
+    request = dict(
+        method="GET",
+        url=httpserver.url_for("/my/path"),
+    )
+
+    with when_a_request_is_being_sent_to_the_server(request):
+
+        with pytest.raises(AssertionError) as exc:
+            httpserver.assert_request(uri="/not/my/path/")
+
+        assert "/not/my/path/" in str(exc)
+        assert "does not match" in str(exc)
+
+
+def test_raises_assertion_error_when_request_was_not_sent(httpserver: BlockingHttpServer):
+    with pytest.raises(AssertionError) as exc:
+        httpserver.assert_request(uri="/my/path/", timeout=1)
+
+    assert "/my/path/" in str(exc)
+    assert "timed out" in str(exc)
+
+
+def test_ignores_when_request_is_not_asserted(httpserver: BlockingHttpServer):
+    request = dict(
+        method="GET",
+        url=httpserver.url_for("/my/path"),
+    )
+
+    with when_a_request_is_being_sent_to_the_server(request) as server_connection:
+
+        assert server_connection.get(timeout=9).text == "No handler found for this request"
+
+
+def test_raises_assertion_error_when_request_was_not_responded(httpserver: BlockingHttpServer):
+    request = dict(
+        method="GET",
+        url=httpserver.url_for("/my/path"),
+    )
+
+    with when_a_request_is_being_sent_to_the_server(request):
+
+        then_the_server_gets_the_request(httpserver, request)
+
+        httpserver.stop()  # waiting for timeout of waiting for the response
+
+        with pytest.raises(AssertionError) as exc:
+            httpserver.check_assertions()
+
+        assert "/my/path" in str(exc)
+        assert "no response" in str(exc).lower()
+
+
+@contextmanager
+def when_a_request_is_being_sent_to_the_server(request):
```

Comment:
I would like these function to be defined at the top of the file, before the tests. Could you please move these to the top?

Functions defined in lines 95, 100, 107, 112, 117 (basically all functions below 95, including line 95).

Reply:
I also see that you are using it from behave, and I have no problems with these namings, but I think I'll try to create a test which does not calls these but have a clearer (to my eyes) way to show how this blockinghttpserver works.

Reply:
I moved them.

Note: I prefer to see the most important things at the beginning of the file and the conventional function definition order is also similar as my original order of functions.

Some helper function gives hint how to implement a test step with the functionality referred by the name of the helper function.